hail/src/main/scala/is/hail/expr/ir/lowering/LoweringPipeline.scala

```diff
@@ -12,12 +12,16 @@ case class LoweringPipeline(lowerings: IndexedSeq[LoweringPass]) {
   final def apply(ctx: ExecuteContext, ir: BaseIR, optimize: Boolean): BaseIR = {
     var x = ir
 
-    if (optimize)
+    TypeCheck.withContext(x, "initial check")
+
+    if (optimize) {
       x = Optimize(x, noisy = true, context = "Lowerer, initial IR", ctx)
+    }
 
     lowerings.foreach { l =>
       try {
         x = l.apply(ctx, x)
+        TypeCheck.withContext(x, l.context)
```

Comment:
Do we want to do this once for each pass, after the optional Optimize? Doesn't give us the granularity we necessarily want, but I'm thinking maybe for non-testing purposes the most important typecheck is right before execution?